src/rust/engine/process_execution/src/remote.rs

```diff
@@ -433,6 +435,107 @@ impl CommandRunner {
+  ) -> BoxFuture<Digest, ExecutionError> {
+    let mut futures = vec![];
+    let path_map = Arc::new(Mutex::new(HashMap::new()));
+    let path_map_2 = path_map.clone();
+    let path_stats: Vec<PathStat> = execute_response
+      .get_result()
+      .get_output_files()
+      .into_iter()
+      .map(|output_file| {
+        let output_file_path_buf = PathBuf::from(output_file.get_path());
+        if output_file.has_digest() {
+          let mut underlying_path_map = path_map.lock().unwrap();
+          underlying_path_map.insert(
+            output_file_path_buf.clone(),
+            output_file.get_digest().into(),
+          );
+          PathStat::file(
+            output_file_path_buf.clone(),
+            File {
+              path: output_file_path_buf.clone(),
+              is_executable: output_file.get_is_executable(),
+            },
+          )
+        } else {
+          let raw_content = output_file.get_content().into();
+          let path_map_3 = path_map.clone();
+          let output_file_path_buf_2 = output_file_path_buf.clone();
+          let output_file_path_buf_3 = output_file_path_buf_2.clone();
+          futures.push(
+            self
+              .store
+              .store_file_bytes(raw_content, false)
+              .map_err(move |error| {
+                ExecutionError::Fatal(format!(
+                  "Error storing raw content for output file {:?}: {:?}",
+                  output_file_path_buf_3, error
+                ))
+              })
+              .map(move |digest| {
+                let mut underlying_path_map = path_map_3.lock().unwrap();
+                underlying_path_map.insert(output_file_path_buf_2, digest);
+              }),
+          );
+          PathStat::file(
+            output_file_path_buf.clone(),
+            File {
+              path: output_file_path_buf.clone(),
+              is_executable: output_file.get_is_executable(),
+            },
+          )
+        }
+      })
+      .collect();
+
+    #[derive(Clone)]
+    struct StoreOneOffRemoteDigest {
+      map_of_paths_to_digests: HashMap<PathBuf, Digest>,
+    }
+
+    impl StoreOneOffRemoteDigest {
+      pub fn new(map: HashMap<PathBuf, Digest>) -> StoreOneOffRemoteDigest {
+        StoreOneOffRemoteDigest {
+          map_of_paths_to_digests: map,
+        }
+      }
+    }
+
+    impl fs::StoreFileByDigest<String> for StoreOneOffRemoteDigest {
+      fn store_by_digest(&self, file: File) -> BoxFuture<Digest, String> {
+        match self.map_of_paths_to_digests.get(&file.path) {
+          Some(digest) => future::ok(digest.clone()),
+          None => future::err(format!(
+            "Error when trying to find digest for path {:?}",
+            file.path
+          )),
+        }.to_boxed()
+      }
+    }
+
+    let store = self.store.clone();
+    future::join_all(futures)
+      .and_then(|_| {
+        let path_wrap_mutex = Arc::try_unwrap(path_map_2).unwrap();
```

Comment:
Throw in a comment to justify that this unwrap is safe (because we've joined on all of the futures that had references to the Arc)